### PR TITLE
Fix fake cluster for empty version

### PR DESCRIPTION
**Pull request by yiminc**
**What changed?**
Correctly handle FakeClusterForEmptyVersion and added integration test for ns migration

**Why?**
Added integration test for migration.

**How did you test it?**
New integration test

**Potential risks**
No

**Is hotfix candidate?**
No

### CODE REVIEW

File: service/history/timerQueueProcessor.go [hunk 1]
```diff
@@ -164,10 +164,7 @@ func (t *timerQueueProcessorImpl) NotifyNewTimers(
 	timerTasks []tasks.Task,
 ) {
 
-	if clusterName == cluster.FakeClusterForEmptyVersion {
-		return
-	}
-	if clusterName == t.currentClusterName {
+	if clusterName == t.currentClusterName || clusterName == cluster.FakeClusterForEmptyVersion {
```
Comment (wxing1292) [inline service/history/timerQueueProcessor.go:167]: `clusterName == cluster.FakeClusterForEmptyVersion`

there should be a dedicated passive task processing logic for this right?
so this is not for active

File: common/cluster/metadata.go [hunk 2]
```diff
@@ -450,7 +459,8 @@ func (m *metadataImpl) updateFailoverVersionToClusterName() {
 func updateVersionToClusterName(clusterInfo map[string]ClusterInformation, failoverVersionIncrement int64) map[int64]string {
 	versionToClusterName := make(map[int64]string)
 	for clusterName, info := range clusterInfo {
-		if failoverVersionIncrement <= info.InitialFailoverVersion || info.InitialFailoverVersion <= 0 {
+		if clusterName != FakeClusterForEmptyVersion &&
+			(failoverVersionIncrement <= info.InitialFailoverVersion || info.InitialFailoverVersion <= 0) {
```
Comment (wxing1292) [inline common/cluster/metadata.go:463]: how about keeping the existing logic and add the `copyClusterInfo[FakeClusterForEmptyVersion] = fakeClusterInfo` to the very end?

Comment (wxing1292) [reply common/cluster/metadata.go:463]: e.g. validate as is, then add the additional FakeClusterForEmptyVersion afterwards

File: service/history/workflow/context.go [hunk 3]
```diff
@@ -520,6 +520,7 @@ func (c *ContextImpl) ConflictResolveWorkflowExecution(
 		newWorkflowEventsSeq,
 		currentWorkflow,
 		currentWorkflowEventsSeq,
+		resetMutableState.GetNamespaceEntry().IsGlobalNamespace(),
```
Comment (wxing1292) [inline service/history/workflow/context.go:523]: can you plz create an issue (ok to assign to me) to use the create workflow execution from transaction.go?

File: service/history/workflowResetter.go [hunk 4]
```diff
@@ -365,6 +365,7 @@ func (r *workflowResetterImpl) persistToDB(
 			currentWorkflowEventsSeq,
 			resetWorkflowSnapshot,
 			resetWorkflowEventsSeq,
+			currentWorkflow.getMutableState().GetNamespaceEntry().IsGlobalNamespace(),
```
Comment (wxing1292) [inline service/history/workflowResetter.go:368]: currentWorkflow -> resetWorkflow

**Review by wxing1292 — APPROVED** [review]
plz check add additional logic in
`clientBean.GetRemoteAdminClient(clusterName)`
this should return an noop client for version 0 cluster

File: service/history/workflowResetter.go (hunk 4 above)
Comment (paulnpdev) [reply service/history/workflowResetter.go:368]: given that the tests were passing without this correction, doesn't that mean that we don't have test that validates this functionality?

File: common/cluster/metadata.go (hunk 2 above)
Comment (yux0) [reply common/cluster/metadata.go:463]: The versionToClusterName uses in ClusterNameForFailoverVersion(), which already been handled. You can just ignore the fake cluster

File: common/cluster/metadata.go [hunk 5]
```diff
@@ -49,6 +49,12 @@ const (
 	FakeClusterForEmptyVersion = "fake-cluster-for-empty-version"
 )
 
+var fakeClusterInfo = ClusterInformation{
+	Enabled:                true,
+	InitialFailoverVersion: 0,
```
Comment (yux0) [inline common/cluster/metadata.go:54]: common.EmptyVersion

File: client/admin/noopClient.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+func (n *noopClient) ListVisibilityTasks(ctx context.Context, in *adminservice.ListVisibilityTasksRequest, opts ...grpc.CallOption) (*adminservice.ListVisibilityTasksResponse, error) {
+	return nil, errFakeNotFound
+}
+
+func (n *noopClient) RemoveTask(ctx context.Context, in *adminservice.RemoveTaskRequest, opts ...grpc.CallOption) (*adminservice.RemoveTaskResponse, error) {
+	return nil, errFakeNotFound
+}
+
+func (n *noopClient) GetWorkflowExecutionRawHistoryV2(ctx context.Context, in *adminservice.GetWorkflowExecutionRawHistoryV2Request, opts ...grpc.CallOption) (*adminservice.GetWorkflowExecutionRawHistoryV2Response, error) {
+	return nil, errFakeNotFound
```
Comment (yux0) [inline client/admin/noopClient.go:84]: The queue executor will check empty version and return retry error. https://github.com/temporalio/temporal/blob/master/service/history/transferQueueStandbyTaskExecutor.go#L518

File: common/cluster/metadata.go [hunk 7]
```diff
@@ -49,6 +49,12 @@ const (
 	FakeClusterForEmptyVersion = "fake-cluster-for-empty-version"
 )
 
+var fakeClusterInfo = ClusterInformation{
+	Enabled:                true,
```
Comment (yux0) [inline common/cluster/metadata.go:53]: IMHO, this should be false and no task notification on fake cluster

**Review by yux0 — APPROVED** [review]

**Pull request merged**